### PR TITLE
Add the hybrid sharding strategy in FSDP (HSDP)

### PR DESCRIPTION
**Pull request by MartinGleize**
**What does this PR do? Please describe:**
Add the hybrid sharding strategy in FSDP (HSDP).
Integration in recipes is only done and tested on instruction finetune, for now.

Fixes #{issue number}

**Does your PR introduce any breaking changes? If yes, please list them:**
No

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**?
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/gang.py [hunk 1]
```diff
@@ -707,6 +707,113 @@ def fake_gangs(device: Device) -> Gangs:
     return Gangs(gang, gang, gang)
 
 
+def setup_2D_mesh_gangs(
```
Comment (cbalioglu) [inline src/fairseq2/gang.py:710]: My preference would be to make this an internal function (i.e. `_setup_2d_mesh_gangs()`). We already expose the actual functionality in `setup_parallel_gangs()` and `to_fsdp()` that are higher-level and abstract away the details of the mesh creation.

File: src/fairseq2/gang.py [hunk 2]
```diff
@@ -707,6 +707,113 @@ def fake_gangs(device: Device) -> Gangs:
...
+        If ``True``, create an underlying ``dist.ProcessGroup`` even for single-rank gangs.
+        The gang is faked otherwise.
+    :param dim_descriptions:
+        String descriptions of returned gangs, used in log and error messages.
+
+    :returns:
+        A ``dict`` of two gangs; 0 maps to the gang of 2D mesh row,
+        1 maps to the gang of the 2D mesh column.
+    """
+    if row_length <= 0:
```
Comment (cbalioglu) [inline src/fairseq2/gang.py:752]: If we make this function internal (as I mentioned above), we can ideally move argument checks to the public functions (`setup_parallel_gangs` and `to_fsdp` with more descriptive error messages. For instance, if the user calls `setup_parallel_gangs` with a wrong `tp_size`, the error message says "`row_length` must be less than or equal to root_gang.size" which is totally clear now.

File: src/fairseq2/nn/fsdp.py [hunk 3]
```diff
@@ -89,10 +90,19 @@ def to_fsdp(
...
+        if local_world_size < 1:
             raise ValueError(
-                f"`local_world_size` must be greater than 0, but is {local_world_size} instead."
+                f"`local_world_size` must be greater than 1, but is {local_world_size} instead."
+            )
+
+        if local_world_size == 1:
+            raise ValueError(
+                f"`local_world_size` must be greater than 1, but is {local_world_size} instead. "
+                "This hybrid configuration would force FSDP to switch to use `NO_SHARD`, "
```
Comment (cbalioglu) [inline src/fairseq2/nn/fsdp.py:104]: Nit. As a convention, we do not line-break individual sentences of an error and log message within the code. This makes it possible for users to "grep" any part of the sentence (e.g. if they see it in the log file and want to figure out where it originates in the code base). This was mostly inspired by Linux kernel coding conventions.

File: tests/unit/test_gang.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            # mesh for 2 hosts, 2 GPUs each:
+            # Host 0: g0 | g1
+            # Host 1: g2 | g3
+            (0, 4, 2, ([0, 1], [0, 2])),
+            (1, 4, 2, ([0, 1], [1, 3])),
+            (2, 4, 2, ([2, 3], [0, 2])),
+            (0, 8, 4, ([0, 1, 2, 3], [0, 4])),
+        ],
+    )
+    def test_setup_2D_mesh_works(
```
Comment (cbalioglu) [inline tests/unit/test_gang.py:94]: Nice! Thanks for writing tests! :)

**Review by cbalioglu — COMMENTED** [review]
Looks quite good. I just have one comment regarding the visibility of `setup_2d_mesh_gangs`.

File: src/fairseq2/gang.py (hunk 1 above)
Comment (MartinGleize) [reply src/fairseq2/gang.py]: I implemented this but added an extra "setup_" function in `gang.py` to keep it coherent with `setup_parallel_gangs()`.

File: src/fairseq2/gang.py (hunk 2 above)
Comment (MartinGleize) [reply src/fairseq2/gang.py]: Good point, less defensive programming in the internal function now.

File: src/fairseq2/nn/fsdp.py (hunk 3 above)
Comment (MartinGleize) [reply src/fairseq2/nn/fsdp.py]: Ah I actually was wondering about that, thank you for the explanation, implemented this.

**Pull request merged**